code/datums/extensions/health/damage_sources.dm

```diff
@@ -0,0 +1,51 @@
+/**
+ * Variant of the health extension that tracks individual damage types.
+ * Separated due to list usage.
+ */
+/datum/extension/health/damage_sources
+	/// List of damage divided into each damage type.
+	var/list/damage_sources = list()
+
+
+/datum/extension/health/damage_sources/New(datum/holder)
+	. = ..()
+	for (var/damage_type in DAMAGE_ALL)
+		damage_sources[damage_type] = 0
+
+
+/datum/extension/health/damage_sources/current_damage(damage_type)
+	return damage_sources[damage_type]
+
+
+/datum/extension/health/damage_sources/mod_health(health_mod, damage_type)
+	. = ..()
+	damage_sources[damage_type] = max(damage_sources[damage_type] - health_mod, 0)
+
+
+/datum/extension/health/damage_sources/revive()
+	. = ..()
+	for (var/damage_type in damage_sources)
+		damage_sources[damage_type] = 0
+
+
+/datum/extension/health/damage_sources/set_max_health(new_max_health)
+	. = ..()
+	return recalculate_health()
```

Comment:
Why are you setting `.` to parent if you're explicitly returning something else?

Reply:
VSC's autocomplete does that and I tend to forget to update it.